accessEntity/data.go

```diff
@@ -0,0 +1,751 @@
+	case IdentifierAppId:
+		return d.getAppCredentialByAppId(priVal, org)
+	}
+	return
+}
+
+func (d *DbManager) GetDevelopers(org, priKey, priVal, secKey, secVal string) (developers []common.Developer, err error) {
+	switch priKey {
+	case IdentifierAppId:
+		return d.getDeveloperByAppId(priVal, org)
+	case IdentifierDeveloperEmail:
+		return d.getDeveloperByEmail(priVal, org)
+	case IdentifierConsumerKey:
+		return d.getDeveloperByConsumerKey(priVal, org)
+	case IdentifierDeveloperId:
+		return d.getDeveloperById(priVal, org)
+	}
+	return
+}
+
+func (d *DbManager) getApiProductsByName(apiProdName string, org string) (apiProducts []common.ApiProduct, err error) {
+	err = d.GetDb().QueryStructs(&apiProducts,
+		sql_select_api_product+
+			`WHERE ap.name = ? AND ap.tenant_id IN `+
+			sql_select_tenant_org,
+		apiProdName,
+		org,
+	)
+	return
+}
+
+func (d *DbManager) getApiProductsByAppId(appId string, org string) (apiProducts []common.ApiProduct, err error) {
+	cols := []string{"*"}
+	query := selectApiProductsById(
+		selectAppCredentialMapperByAppId(
+			"'"+appId+"'",
+			"apiprdt_id",
+		),
+		cols...,
+	) + " AND ap.tenant_id IN " + sql_select_tenant_org
+	//log.Debugf("getApiProductsByAppId: %v", query)
+	err = d.GetDb().QueryStructs(&apiProducts, query, org)
+	return
+}
+
+func (d *DbManager) getApiProductsByConsumerKey(consumerKey string, org string) (apiProducts []common.ApiProduct, err error) {
+	cols := []string{"*"}
+	query := selectApiProductsById(
+		selectAppCredentialMapperByConsumerKey(
+			"'"+consumerKey+"'",
+			"apiprdt_id",
+		),
+		cols...,
+	) + " AND ap.tenant_id IN " + sql_select_tenant_org
+	//log.Debugf("getApiProductsByConsumerKey: %v", query)
+	err = d.GetDb().QueryStructs(&apiProducts, query, org)
+	return
+}
+
+func (d *DbManager) getApiProductsByAppName(appName, devEmail, devId, comName, org string) (apiProducts []common.ApiProduct, err error) {
+	cols := []string{"*"}
+	var appQuery string
+	switch {
+	case devEmail != "":
+		appQuery = selectAppByNameAndDeveloperId(
+			"'"+appName+"'",
+			selectDeveloperByEmail(
+				"'"+devEmail+"'",
+				"id",
+			),
+			"id",
+		)
+	case devId != "":
+		appQuery = selectAppByNameAndDeveloperId(
+			"'"+appName+"'",
+			"'"+devId+"'",
+			"id",
+		)
+	case comName != "":
+		appQuery = selectAppByNameAndCompanyId(
+			"'"+appName+"'",
+			selectCompanyByName(
+				"'"+comName+"'",
+				"id",
+			),
+			"id",
+		)
+	default:
+		appQuery = selectAppByName(
+			"'"+appName+"'",
+			"id",
+		)
+	}
+
+	query := selectApiProductsById(
+		selectAppCredentialMapperByAppId(
+			appQuery,
+			"apiprdt_id",
+		),
+		cols...,
+	) + " AND ap.tenant_id IN " + sql_select_tenant_org
+	//log.Debugf("getApiProductsByAppName: %v", query)
+	err = d.GetDb().QueryStructs(&apiProducts, query, org)
+	return
+}
+
+func (d *DbManager) getAppByAppId(id, org string) (apps []common.App, err error) {
+	cols := []string{"*"}
+	query := selectAppById(
+		"'"+id+"'",
+		cols...,
```

Comment:
Looks fancy, what is this doing and why this way?

Reply:
The motivation was to reuse code, and increase extensibility.
In fact, each access entity request has different logics according to identifiers. Sometimes there're 2 or 3 table joins happenning.
A straightforward way is to hard-code all nested long SQL queries for each entity-primaryIdentifier-secondaryIdentifier combination. 
However, if there's any change in db schema, all affected SQL queries have to be changed. Hard will also a non-extensible manual process. 
We discussed about the `http://graphql.org/` during API design, but that would be too much work for MP. So we decided to use normal http GET for accessEntity. However, ideally in the future this will be the only portal to access KMS data, and there will be a set of generic apis to do that. So I designed it in this way to enable it generate SQL statements at runtime. The generic logics & extensible design are on our roadmap to a final generic plugin that exposes KMS data.
Any suggestions or concerns?